operator/pkg/ingress/ingress_reconcile.go

```diff
@@ -0,0 +1,428 @@
+	}
+
+	for k, v := range resources {
+		if err := r.tryDeletingResource(ctx, k, v); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ingressReconciler) tryCleanupSharedResources(ctx context.Context) error {
+	// In shared loadbalancing mode, only the CiliumEnvoyConfig is managed by the Operator.
+	// Service and Endpoints are created by the Helm Chart.
+	desiredCiliumEnvoyConfig, err := r.buildSharedResources(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to build shared resources: %w", err)
+	}
+
+	if err := r.createOrUpdateCiliumEnvoyConfig(ctx, desiredCiliumEnvoyConfig); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ingressReconciler) buildSharedResources(ctx context.Context) (*ciliumv2.CiliumEnvoyConfig, error) {
+	ingressList := networkingv1.IngressList{}
+	if err := r.client.List(ctx, &ingressList); err != nil {
+		return nil, fmt.Errorf("failed to list Ingresses: %w", err)
+	}
+
+	m := &model.Model{}
+	for _, item := range ingressList.Items {
+		if !isCiliumManagedIngress(ctx, r.client, r.logger, item) || r.isEffectiveLoadbalancerModeDedicated(&item) || item.GetDeletionTimestamp() != nil {
+			continue
+		}
+		if annotations.GetAnnotationTLSPassthroughEnabled(&item) {
+			m.TLS = append(m.TLS, ingestion.IngressPassthrough(item, r.defaultSecretNamespace, r.defaultSecretName)...)
+		} else {
+			m.HTTP = append(m.HTTP, ingestion.Ingress(item, r.defaultSecretNamespace, r.defaultSecretName)...)
+		}
+	}
+
+	cec, _, _, err := r.sharedTranslator.Translate(m)
+
+	return cec, err
+}
+
+func (r *ingressReconciler) buildDedicatedResources(ctx context.Context, ingress *networkingv1.Ingress) (*ciliumv2.CiliumEnvoyConfig, *corev1.Service, *corev1.Endpoints, error) {
+	m := &model.Model{}
+
+	if annotations.GetAnnotationTLSPassthroughEnabled(ingress) {
+		m.TLS = append(m.TLS, ingestion.IngressPassthrough(*ingress, r.defaultSecretNamespace, r.defaultSecretName)...)
+	} else {
+		m.HTTP = append(m.HTTP, ingestion.Ingress(*ingress, r.defaultSecretNamespace, r.defaultSecretName)...)
+	}
+
+	cec, svc, ep, err := r.dedicatedTranslator.Translate(m)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to translate model into resources: %w", err)
+	}
+
+	r.propagateIngressAnnotationsAndLabels(ingress, &svc.ObjectMeta)
+
+	// Explicitly set the controlling OwnerReference on the CiliumEnvoyConfig
+	if err := controllerutil.SetControllerReference(ingress, cec, r.client.Scheme()); err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to set controller reference on CiliumEnvoyConfig: %w", err)
+	}
+
+	return cec, svc, ep, err
+}
+
+func (r *ingressReconciler) createOrUpdateCiliumEnvoyConfig(ctx context.Context, desiredCEC *ciliumv2.CiliumEnvoyConfig) error {
+	cec := desiredCEC.DeepCopy()
+
+	result, err := controllerutil.CreateOrUpdate(ctx, r.client, cec, func() error {
```

Comment:
Any reason to use CreateOrUpdate but not CreateOrPatch here?

Reply:
> Any reason to use CreateOrUpdate but not CreateOrPatch here?

No specific reason. 

This and potential further optimization of unnecessary writes could be part of a follow up. As discussed - this would also involve optimizing the translation logic. E.g. we should build the desired state on top of a resource reference that is potentially prefilled with the current loaded state. This way we could really rely on the DeepEqual check when checking whether we should update/patch the resource.